torcheval/metrics/functional/statistical/wasserstein.py

```diff
@@ -0,0 +1,185 @@
+    y_weights: Optional[torch.Tensor] = None,
+) -> torch.Tensor:
+    r"""
+    The Wasserstein distance, also called the Earth Mover's Distance, is a
+    measure of the similarity between two distributions.
+
+    The Wasserstein distance between two distributions is intuitively the
+    minimum weight of soil (times distance moved) that would need to be moved
+    if the two distributions were represented by two piles of soil.
+
+    Args
+    ----------
+    x, y (Tensor) : 1D Tensor values observed in the distribution.
+    x_weights, y_weights (Tensor): Optional tensor weights for each value.
+        If unspecified, each value is assigned the same value.
+        `x_weights` (resp. `y_weights`) must have the same length as
+        `x` (resp. `y`). If the weight sum differs from 1, it
+        must still be positive and finite so that the weights can be normalized
+        to sum to 1.
+
+    Returns
+    -------
+    distance : Tensor value
+        The computed distance between the distributions.
+
+    Notes
+    -----
+    The first Wasserstein distance between the distributions :math:`x` and
+    :math:`x` is:
+
+    .. math::
+
+        W_1 (x, y) = \inf_{\pi \in \Gamma (x, y)} \int_{\mathbb{R} \times
+        \mathbb{R}} |p-q| \mathrm{d} \pi (p, q)
+
+    where :math:`\Gamma (x, y)` is the set of (probability) distributions on
+    :math:`\mathbb{R} \times \mathbb{R}` whose marginals are :math:`x` and
+    :math:`y` on the first and second factors respectively.
+
+    If :math:`X` and :math:`Y` are the respective CDFs of :math:`x` and
+    :math:`y`, this distance also equals to:
+
+    .. math::
+
+        W_1(x, y) = \int_{-\infty}^{+\infty} |X-Y|
+
+    See [2]_ for a proof of the equivalence of both definitions.
+
+    The input distributions can be empirical, therefore coming from samples
+    whose values are effectively inputs of the function, or they can be seen as
+    generalized functions, in which case they are weighted sums of Dirac delta
+    functions located at the specified values.
+
+    References
+    ----------
+    .. [1] "Wasserstein metric", https://en.wikipedia.org/wiki/Wasserstein_metric
+    .. [2] Ramdas, Garcia, Cuturi "On Wasserstein Two Sample Testing and Related
+           Families of Nonparametric Tests" (2015). :arXiv:`1509.02237`.
+
+    Examples
+    --------
+    >>> from torcheval.metrics.functional import wasserstein_1d
+    >>> wasserstein_1d(torch.tensor([0,1,2]), torch.tensor([0,1,1]))
+    torch.tensor([0.33333333333333337])
+    >>> wasserstein_1d(torch.tensor([0,1,2]), torch.tensor([0,1,1]), torch.tensor([1,2,0]), torch.tensor([1,1,1]))
+    torch.tensor([0.0])
+    >>> wasserstein_1d(torch.tensor([0,1,2,2]), torch.tensor([0,1]))
+    torch.tensor([0.75])
+
+    """
+    _wasserstein_param_check(x, y, x_weights, y_weights)
+    _wasserstein_update_input_check(x, y, x_weights, y_weights)
+    return _wasserstein_compute(x, y, x_weights, y_weights)
+
+
+def _wasserstein_param_check(
```

Comment:
small refactor here, `params` for us are things that don't change between calls to `update`, e.g. constructor args. It seems the checks in `param_check` should be in `update_input_check`. It looks like you don't have any params, so you can just delete this function and add the checks to `update_input_check`.

Reply:
Thank you for this thorough review @bobakfb ! I will try to get this done over the weekend.

Reply:
This change has been implemented, thanks!